### PR TITLE
Prune leaves on finality

### LINKED ISSUE #1558: Prune DB leaf set on finality
Currently all leaves on the block tree are kept; old ones which are below the `finalized_number` and are not canonical should be pruned.

### PR DESCRIPTION
**Pull request by rphmeier**
Otherwise the leaves set grows unbounded.

Closes #1558

### CODE REVIEW

File: core/client/db/src/lib.rs [hunk 1]
```diff
@@ -892,9 +908,16 @@ impl<Block: BlockT<Hash=H256>> Backend<Block> {
 
 		if let Some((number, hash, enacted, retracted, displaced_leaf, is_best)) = imported {
 			if let Err(e) = write_result {
+				let mut leaves = self.blockchain.leaves.write();
```
Comment (gterzian) [inline core/client/db/src/lib.rs]: Could it be better to hold the lock for a little as possible, like in the below?

```
if let Some(displaced_leaf) = displaced_leaf {
					self.blockchain.leaves.write().undo().undo_import(displaced_leaf);
				}

				if let Some(finalization_displaced) = finalization_displaced_leaves {
					self.blockchain.leaves.write().undo().undo_finalization(finalization_displaced);
				}
```

Unless off-course you intend to exclude any reading or writing that might occur between the two method calls...

Comment (rphmeier) [reply core/client/db/src/lib.rs]: `try_commit_operation` is meant to be synchronous as part of the API. It should only race with read calls, e.g. to `blockchain::Backend::leaves`. In that case, I think it's better either to give the leaves before, but not after.

Comment (rphmeier) [reply core/client/db/src/lib.rs]: I'll file a broader issue about refactoring the `LeafSet` to use a transaction -> update pattern. https://github.com/paritytech/substrate/issues/1878

**Comment by rphmeier** [comment]
@marcio-diaz I think it's fine for the children map to continue growing as long as it does so on-disk.

**Pull request merged**